### PR TITLE
- Added variant filtering option for the file tree

### PR DESCRIPTION
**Pull request by RustyClaymore**

### CODE REVIEW

File: src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx [hunk 1]
```diff
@@ -87,4 +113,41 @@ export class Sidebar extends React.Component<{}, {}> {
     private openFolder = () => {
         openFolderExternally();
     };
+
+    private onVariantFilterSelected = (filter: string) => {
```
Comment (christianblos) [inline src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx]: pls rename this to onSelectVariantFilter. So the naming is consistent

File: src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx [hunk 2]
```diff
@@ -87,4 +113,41 @@ export class Sidebar extends React.Component<{}, {}> {
...
+        this.injected.projectState.project.setFilter(filter);
+    };
+
+    private onResetFilter = () => {
+        this.injected.projectState.project.setFilter(null);
+    };
+
+    private getUniqueVariantIds(): string[] {
```
Comment (christianblos) [inline src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx]: this function can be simplified:
```
private getUniqueVariantIds(): string[] {
    let files = this.injected.projectState.project.schemaTree.getFilesFlat();
    let uniqueVariantIds: Set<string> = new Set();

    for (let file of files) {
        for (let variant of file.variants) {
            uniqueVariantIds.add(variant.variantId);
        }
    }

    return [...uniqueVariantIds];
}
```

File: src/app/renderer/components/common/IconDropdown/IconDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import * as React from "react";
+import {If} from "../../helper/If";
+import styles from "./IconDropdownStyles.scss"
+import {Link} from "../Link";
+import {Icon} from "../Icon";
+
+interface Props {
+    optionsList?: string[];
```
Comment (christianblos) [inline src/app/renderer/components/common/IconDropdown/IconDropdown.tsx]: optionsList is not used

File: src/domain/states/objects/fileTree/__tests__/SchemaTree.test.ts [hunk 4]
```diff
@@ -62,4 +63,56 @@ describe('SchemaTree', () => {
...
+            ]),
+            new SchemaDir('Dir 2', 'dir2', [
+                fileC,
+            ]),
+            fileD,
+        ]);
+
+        let expectedFlatTree = [fileA, fileB, fileC, fileD];
+
+        expect([...new Set(tree.getFilesFlat())]).toEqual(expectedFlatTree);
```
Comment (christianblos) [inline src/domain/states/objects/fileTree/__tests__/SchemaTree.test.ts]: why do you use a Set here? I think you should test without a Set. If getFlatFiles() returns one file multiple times, then it's wrong

Comment (christianblos) [reply src/domain/states/objects/fileTree/__tests__/SchemaTree.test.ts]: I see the problem. SchemaTree.reduceFiles should look like this:
```
if (child instanceof SchemaFile) {
    return [...value, child];
} else {
    return (child as SchemaDir).children.reduce(this.reduceFiles, value);
}
```

And please update the test to don't use the Set:
```
expect(tree.getFilesFlat()).toEqual(expectedFlatTree);
```

File: src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx (hunk 1 above)
Comment (RustyClaymore) [reply src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx]: Done

File: src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx (hunk 2 above)
Comment (RustyClaymore) [reply src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx]: done

File: src/app/renderer/components/app/FileBrowser/components/Sidebar/SidebarStyles.scss [hunk 5]
```diff
@@ -7,4 +7,46 @@
...
+
+    a:hover {
+      background-color: $color-bg-highlight;
+    }
+  }
+
+  .filter {
+    display: block;
+    position: relative;
+    background-color: $color-primary;
```
[the remark this answers is not in the document]
Comment (RustyClaymore) [reply src/app/renderer/components/app/FileBrowser/components/Sidebar/SidebarStyles.scss]: done

File: src/app/renderer/components/common/IconDropdown/IconDropdown.tsx (hunk 3 above)
Comment (RustyClaymore) [reply src/app/renderer/components/common/IconDropdown/IconDropdown.tsx]: done

File: src/app/renderer/components/common/IconDropdown/IconDropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+interface State {
+    isOpen: boolean;
+}
+
+export class IconDropdown extends React.Component<Props, State> {
+    constructor(props: Props) {
+        super(props);
+        
+        this.props = {
```
[the remark this answers is not in the document]
Comment (RustyClaymore) [reply src/app/renderer/components/common/IconDropdown/IconDropdown.tsx]: done

File: src/app/renderer/components/common/IconDropdown/IconDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+            icon: ''
+        };
+
+        this.state = {
+            isOpen: false,
+        };
+
+        this.onOpen = this.onOpen.bind(this);
+        this.onClose = this.onClose.bind(this);
```
[the remark this answers is not in the document]
Comment (RustyClaymore) [reply src/app/renderer/components/common/IconDropdown/IconDropdown.tsx]: done

File: src/app/renderer/components/common/IconDropdown/IconDropdown.tsx [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    render() {
+        return (
+            <>
+                <Link title='Filter' onClick={this.onOpen}><Icon value={this.props.icon}/></Link>
```
[the remark this answers is not in the document]
Comment (RustyClaymore) [reply src/app/renderer/components/common/IconDropdown/IconDropdown.tsx]: done

File: src/app/renderer/components/common/IconDropdown/IconDropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+                <If cond={this.state.isOpen}>
+                    <div className={styles.dropdown}>
+                        {this.props.children}
+                    </div>
+                </If>
+            </>
+        );
+    };
+
+    public onClose = () => {
```
[the remark this answers is not in the document]
Comment (RustyClaymore) [reply src/app/renderer/components/common/IconDropdown/IconDropdown.tsx]: nice!

File: src/domain/states/objects/fileTree/__tests__/SchemaTree.test.ts (hunk 4 above)
Comment (RustyClaymore) [reply src/domain/states/objects/fileTree/__tests__/SchemaTree.test.ts]: done

File: src/app/renderer/components/common/IconDropdown/IconDropdownStyles.scss [hunk 10]
```diff
@@ -0,0 +1,8 @@
+@import "../../../config/config";
+
+:local .dropdown{
```
[the remark this answers is not in the document]
Comment (RustyClaymore) [reply src/app/renderer/components/common/IconDropdown/IconDropdownStyles.scss]: It's used for the dropdown background, but removed the redundancy from SidebarStyles

File: src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx [hunk 11]
```diff
@@ -37,9 +40,28 @@ export class Sidebar extends React.Component<{}, {}> {
                     <Link title='Expand All' onClick={this.expandAll}><Icon value={Icon.EXPAND}/></Link>
                     <Link title='Collapse All' onClick={this.collapseAll}><Icon value={Icon.COLLAPSE}/></Link>
                     <Link title='Open folder' onClick={this.openFolder}><Icon value={Icon.EXTERNAL_LINK}/></Link>
+                    <IconDropdown icon={Icon.FILTER}>
+                        <div className={styles.dropdown}>
+                            <li className={styles.dropdownContent}>
```
[the remark this answers is not in the document]
Comment (RustyClaymore) [reply src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx]: Done! Added Menu and MenuItem components and added their styles.

File: src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx [hunk 12]
```diff
@@ -37,9 +40,28 @@ export class Sidebar extends React.Component<{}, {}> {
...
+                                {
+                                    this.getUniqueVariantIds().map((variantId: string, index: number) => {
+                                        return (
+                                            <a onClick={() => this.onVariantFilterSelected(variantId)}
+                                               key={index}>{variantId}</a>
```
[the remark this answers is not in the document]
Comment (RustyClaymore) [reply src/app/renderer/components/app/FileBrowser/components/Sidebar/Sidebar.tsx]: done

**Pull request merged**